### PR TITLE
Add OSType Support for AzureManagedNodePools to add Windows Support

### PR DESCRIPTION
**Pull request by luthermonson**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**: Adds the OSType functionality to AKS Managed Clusters which allows you to choose Linux or Windows for your User nodepools, this PR was inspired by the work done by @zmalik and I used this PR as a roadmap: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1925

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes part of #1701

**Special notes for your reviewer**:
Technically you have to be using the Azure CNI network policy but I didn't see any great way in the AzureManagedNodePool validators to grab the cluster to confirm this. If there is a way to do this you'd like me to add I'm more than willing.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
enable AKS Windows node pools
```

### CODE REVIEW

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go [hunk 1]
```diff
@@ -269,6 +274,18 @@ func (m *AzureManagedMachinePool) validateMaxPods() error {
 	return nil
 }
 
+func (m *AzureManagedMachinePool) validateOSType() error {
+	if m.Spec.OSType != nil && *m.Spec.OSType == azure.WindowsOS {
+		if m.Spec.Mode != string(NodePoolModeUser) {
+			return field.Forbidden(
+				field.NewPath("Spec", "OSType"),
+				"Windows OSType requires AzureManagedMachinePool mode User")
```
Comment (jackfrancis) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:282]: I have a slight preference for expressing "System node pool must have OSType 'Linux'" or something similar. And so I'd probably invert the validation logic:

- if mode == system
  - if ostype != linux
    - then return error

**Comment by jackfrancis** [comment]
@luthermonson thanks for this!

When updating the type system you'll always need to run `make generate` afterwards to render generated code. (then commit the changes)

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go (hunk 1 above)
Comment (luthermonson) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go]: fixed

**Comment by luthermonson** [comment]
@jackfrancis i had forgotten the AgentPoolSpec fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/d765eb83b9da8efa4300533c286e35f849469818..469137da9e5ecb305e91dc71fff3996da7e51ff4 and missed two generated fixes: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/469137da9e5ecb305e91dc71fff3996da7e51ff4..4dd5116a4b41064e2d30ebb8fda2ae5c69dd8cc5

i think this is good to go now based on your feedback

File: azure/scope/managedcontrolplane_test.go [hunk 2]
```diff
@@ -267,3 +267,135 @@ func TestManagedControlPlaneScope_AddonProfiles(t *testing.T) {
...
+					},
+				},
+				ManagedMachinePools: []ManagedMachinePool{
+					{
+						MachinePool:      getMachinePool("pool0"),
+						InfraMachinePool: getAzureMachinePool("pool0", infrav1.NodePoolModeSystem),
+					},
+					{
+						MachinePool:      getMachinePool("pool1"),
+						InfraMachinePool: getLinuxAzureMachinePool("pool1"),
```
Comment (jackfrancis) [inline azure/scope/managedcontrolplane_test.go:311]: this helper func `getLinuxAzureMachinePool` doesn't actually exist

**Comment by jackfrancis** [comment]
@luthermonson while testing this out I got this error from the AKS API, so let's add webhook validation for this requirement:

```
E0624 08:36:31.645439       1 controller.go:317] controller/azuremanagedcontrolplane "msg"="Reconciler error" "error"="error creating AzureManagedControlPlane capz-e2e-tdeuea/capz-e2e-tdeuea-aks: failed to reconcile AzureManagedControlPlane service managedcluster: failed to create resource capz-e2e-tdeuea-aks/capz-e2e-tdeuea-aks (service: managedcluster): Code=\"WindowsAgentPoolNameTooLong\" Message=\"Windows agent pool name can not be longer than 6 characters.\"" "name"="capz-e2e-tdeuea-aks" "namespace"="capz-e2e-tdeuea" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureManagedControlPlane"
```

File: azure/scope/managedcontrolplane_test.go (hunk 2 above)
Comment (luthermonson) [reply azure/scope/managedcontrolplane_test.go:311]: ... ya i dont know how that happened. i found them in my local history and will commit

**Comment by luthermonson** [comment]
@jackfrancis this should take care of the remaining issues. thanks https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/4dd5116a4b41064e2d30ebb8fda2ae5c69dd8cc5..6e6edd207791f7ffc464b451a8978dcc9c954062

File: exp/api/v1alpha4/zz_generated.conversion.go [hunk 3]
```diff
@@ -1194,6 +1194,7 @@ func autoConvert_v1beta1_AzureManagedMachinePoolSpec_To_v1alpha4_AzureManagedMac
 	// WARNING: in.MaxPods requires manual conversion: does not exist in peer-type
 	// WARNING: in.OsDiskType requires manual conversion: does not exist in peer-type
 	// WARNING: in.EnableUltraSSD requires manual conversion: does not exist in peer-type
+	// WARNING: in.OSType requires manual conversion: does not exist in peer-type
```
Comment (jsturtevant) [inline exp/api/v1alpha4/zz_generated.conversion.go:1197]: I think the tests are failing since we need to convert this manually. look for `OsDiskType` as an example

Comment (luthermonson) [reply exp/api/v1alpha4/zz_generated.conversion.go:1197]: fixed!

**Comment by jackfrancis** [comment]
@luthermonson PR looks great, let's do one final thing to ensure this lands w/ E2E test coverage:

1. rebase on top of latest main (which includes https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2420)
2. update the cluster template that we use to test AKS scenarios in PRs (identified by the `pull-cluster-api-provider-azure-e2e-exp` test job):
  a. make the `AzureManagedMachinePool` resource in this file a Windows pool: `templates/test/ci/prow-aks/patches/aks-pool2.yaml`
  b. run `hack/gen-flavors.sh` from the git root of this project to render the fully generated template w/ the update

tl;dr your diff should look like this:

```
$ git diff
diff --git a/templates/test/ci/cluster-template-prow-aks.yaml b/templates/test/ci/cluster-template-prow-aks.yaml
index e1888873..d6a17e89 100644
--- a/templates/test/ci/cluster-template-prow-aks.yaml
+++ b/templates/test/ci/cluster-template-prow-aks.yaml
@@ -163,4 +163,5 @@ metadata:
   namespace: default
 spec:
   mode: User
+  osType: Windows
   sku: ${AZURE_NODE_MACHINE_TYPE}
diff --git a/templates/test/ci/prow-aks/patches/aks-pool2.yaml b/templates/test/ci/prow-aks/patches/aks-pool2.yaml
index a7b8d9de..5622a610 100644
--- a/templates/test/ci/prow-aks/patches/aks-pool2.yaml
+++ b/templates/test/ci/prow-aks/patches/aks-pool2.yaml
@@ -26,3 +26,4 @@ metadata:
 spec:
   mode: User
   sku: "${AZURE_NODE_MACHINE_TYPE}"
+  osType: Windows
```

Then we can actually build a AKS cluster w/ a Windows node pool against this PR changeset before merging.

Thanks!

**Comment by jackfrancis** [comment]
/release-note-edit ```release-note enable AKS Windows node pools ```

**Comment by luthermonson** [comment]
@jackfrancis rebased and added this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2410/commits/8b32ae11a903fe8361e72538b01e07ce65ee0cc0 hopefully this is right

**Comment by jackfrancis** [comment]
@luthermonson looking great (and tests passed as well)

can you kindly squash commits and we'll land this bird
